### PR TITLE
SDIT-1619 Locations reconciliation

### PR DESCRIPTION
**Pull request by steverendell**
Tweaks to location history checks

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/locations/LocationsReconciliationService.kt [hunk 1]
```diff
@@ -249,39 +249,38 @@ class LocationsReconciliationService(
...
-          nomisItem.oldValue == dpsItem.oldValue &&
-          nomisItem.newValue == dpsItem.newValue
-        ) {
-          // matches
-        } else {
-          return true
-        }
+    if (dpsSetSorted.size > nomisSetSorted.size) {
+      return true
+    }
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/locations/LocationsReconciliationService.kt:264]: Minor style thing for next time, the Kotlin way would read slightly better (might need reverse logic, e.g NOT equal)

`nomisSetSorted.zip(dpsSetSorted).any { (nomisItem, dpsItem) -> 
       nomisItem.attribute == dpsItem.attribute &&
        nomisItem.amendedBy == dpsItem.amendedBy &&
        nomisItem.amendedDate == dpsItem.amendedDate &&
        nomisItem.oldValue == dpsItem.oldValue &&
        nomisItem.newValue == dpsItem.newValue
      ) 
    }`

**Review by andymarke — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/locations/LocationsReconciliationService.kt (hunk 1 above)
Comment (steverendell) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/locations/LocationsReconciliationService.kt:271]: Good idea - done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/locations/LocationsApiService.kt [hunk 2]
```diff
@@ -5,21 +5,22 @@ import org.springframework.beans.factory.annotation.Qualifier
...
 import uk.gov.justice.digital.hmpps.prisonertonomisupdate.services.typeReference
 
 @Service
 class LocationsApiService(@Qualifier("locationsApiWebClient") private val webClient: WebClient) {
-  suspend fun getLocation(id: String, includeHistory: Boolean = false): Location {
+  suspend fun getLocation(id: String, includeHistory: Boolean = false): LegacyLocation {
     return webClient.get()
       .uri {
-        it.path("/locations/{id}")
+        it.path("/sync/id/{id}")
```
Comment (steverendell) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/locations/LocationsApiService.kt:18]: There is a new locations endpoint which returns the location in a more 'nomis friendly' format so switching to use that

File: src/main/resources/application-localstack.yml [hunk 3]
```diff
@@ -5,27 +5,57 @@ hmpps.sqs:
       queueName: hmpps_prisoner_to_nomis_visit_queue
       dlqName: hmpps_prisoner_to_nomis_visit_dlq
       subscribeTopicId: domainevents
-      subscribeFilter: "{\"eventType\":[\"prison-visit.booked\", \"prison-visit.changed\", \"prison-visit.cancelled\"]}"
+      subscribeFilter: >-
+        { "eventType":[
+        "prison-visit.booked",
+        "prison-visit.changed",
+        "prison-visit.cancelled"]}
```
Comment (steverendell) [inline src/main/resources/application-localstack.yml:12]: I thought Id convert all the remaining filters to the new more readable format

File: src/main/resources/application-localstack.yml [hunk 4]
```diff
@@ -5,27 +5,57 @@ hmpps.sqs:
...
+        { "eventType":[
+        "activities.activity-schedule.created",
+        "activities.activity-schedule.amended",
+        "activities.scheduled-instance.amended",
+        "activities.prisoner.allocated",
+        "activities.prisoner.deallocated",
+        "activities.prisoner.allocation-amended",
+        "activities.prisoner.attendance-created",
+        "activities.prisoner.attendance-amended",
+        "activities.prisoner.attendance-expired"]}
```
Comment (steverendell) [inline src/main/resources/application-localstack.yml:37]: this list was out of date

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/locations/LocationsReconciliationService.kt [hunk 5]
```diff
@@ -90,23 +91,24 @@ class LocationsReconciliationService(
           allDpsIdsInNomis.contains((it.id.toString()))
         }
         .map { dpsRecord ->
+          val legacyRecord = locationsApiService.getLocation(dpsRecord.id.toString(), true)
```
Comment (steverendell) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/locations/LocationsReconciliationService.kt:94]: Now need to get each location in legacy format. I'll talk to Mike W about adding another endpoint that gets pages of locations in legacy so we can avoid this extra call

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/locations/LocationsApiService.kt [hunk 6]
```diff
@@ -5,21 +5,22 @@ import org.springframework.beans.factory.annotation.Qualifier
...
 import org.springframework.web.reactive.function.client.WebClient
+import uk.gov.justice.digital.hmpps.prisonertonomisupdate.locations.model.LegacyLocation
 import uk.gov.justice.digital.hmpps.prisonertonomisupdate.locations.model.Location
 import uk.gov.justice.digital.hmpps.prisonertonomisupdate.services.RestResponsePage
 import uk.gov.justice.digital.hmpps.prisonertonomisupdate.services.typeReference
 
 @Service
 class LocationsApiService(@Qualifier("locationsApiWebClient") private val webClient: WebClient) {
-  suspend fun getLocation(id: String, includeHistory: Boolean = false): Location {
+  suspend fun getLocation(id: String, includeHistory: Boolean = false): LegacyLocation {
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/locations/LocationsApiService.kt:15]: Does Legacy mean NOMIS?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/locations/LocationsReconciliationService.kt [hunk 7]
```diff
@@ -248,45 +254,40 @@ class LocationsReconciliationService(
...
   private fun reportHistory(nomisList: List<AmendmentResponse>?, dpsList: List<ChangeHistory>?): String {
-    val nomisListConverted = nomisList!!.map { nomis ->
+    val nomisListConverted = filterAmendmentResponses(nomisList)?.map { nomis ->
       ChangeHistory(toHistoryAttribute(nomis.columnName), nomis.amendedBy, nomis.amendDateTime, nomis.oldValue, nomis.newValue)
-    }
+    } ?: emptyList()
 
-    val onlyInDps = dpsList!! - nomisListConverted
-    val onlyInNomis = nomisListConverted - dpsList
+    val dpsListNonNull = dpsList ?: emptyList()
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/locations/LocationsReconciliationService.kt:288]: Nullability is part of the Type of this object - not sure you need to includethe type in the name?

**Review by andymarke — APPROVED** [review]
lgtm

**Pull request merged**